internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,123 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(url, clientId, clientSecret string) *LazyToken {
+	return &LazyToken{
+		url:          url,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+	}
+}
```

Comment:
Please create another similar initializer function that allows to create new token with offline token too. Name it appropriately.

Reply:
As I said, instead of one "contructor" functions, I suggest to create two of them:

* `NewLazyToken(url, clientId, clientSecret) *LazyToken`
* `NewOfflineLazyToken(url, clientId, offlineToken string) *LazyToken`